keps/sig-node/20190425-node-readiness-gates.md

```diff
@@ -0,0 +1,389 @@
+}
+
+// NodeReadinessGate indicates a set of pods that must be ready on a given node before the
+// node itself is considered ready. Pods listed under NodeReadinessGates should be system critical pods
+// that should block other workloads from running on a node until it is ready.
+type NodeReadinessGate struct {
+   // The name of the readiness gate, mainly required as the patch merge key
+   Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+   // A label query for pods representing this readiness gate.
+   // Generally, labels here should point to a specific pod on a node.
+   Selector LabelSelector `json:"selector,omitempty" protobuf:"bytes,2,opt,name=selector"`
+   // The namespace query for the pods representing this readiness gate.
+   Namespace string `json:"namespace,omitempty" protobuf:"bytes,3,opt,name=namespace"`
+   // Indicates if this readiness gate should block the overall readiness of a node. Default: true
+   // +optional
+   BlocksReadiness bool `json:"blocksReadiness,omitempty" protobuf:"varint,4,opt,name=blocksReadiness"`
+}
+```
+
+For each `NodeReadinessGate`, kubelet will fetch the state of all pods that match its label selector and take the readiness state of those pods into account.
+If all pods represented by every `NodeReadinessGate` is ready (based on container state and readiness probes) then the node is also considered ready.
+If no pods are found with the label selector then the node is consider to be "Not Ready".
+
+The individual pods representing readiness gates are required to detect kubelet crash (or restart) quickly and reflect that in their own readiness states via probes.
+The kubelet places a special meta readiness taint on a node object if 1 or more of those pods are not ready. If the Node Lifecycle Controller observes that all the pods selected
+by `ReadinessGates` are ready, it will remove that taint immediately.
+
+If `BlocksReadiness` is set to `false`, the overall readiness of a node is not blocked by the pods under this specific readiness gate. This is useful for non-critical pods
+that should not block the readiness of a node but should be reported in the node status.
+
+### Readiness Gates from External Sources
+
+The `ReadinessGates` field can be used by external components to dictate node readiness since they can apply an arbitrary readiness gate with a reserved label selector
+that should not be used by any pods on the cluster. A common usecase here is an external cloud provider that wants to ensure no pods are scheduled on a node until
+networking for a given node is properly set up by the cloud provider. In this case, the cloud provider would apply any custom readiness gates to a node during registration
+and a separate controller would remove the readiness gate when it sees that node's networking has been properly configured. No pods matching the label selector of that
```

Comment:
> And if it has permission to modify the Node object anyway, couldn't it just apply and then remove a taint directly rather than fiddling with the readiness gates?

There's time between when the cloud provider can apply that initial taint and when the node is initialized. It's true that a cloud provider can apply custom taints to achieve the same behavior but it will always race against the kubelet for pods unless we add the gate during initialization.

Reply:
> This is kind of totally a gross hack. The cloud provider tells the node "I won't be ready until event X happens", knowing full well that event X will never happen, and then just says "never mind, forget I said that" when it's ready. It's not using the defined mechanism, it's abusing it.

I have mixed feelings about this... yes it's not exactly using the feature the way it is fully intended but it satisfies _most_ of the use-cases around readiness - which is waiting for system critical pods (usually DaemonSets) - while also allowing for the external cloud provider case.

Reply:
I'm not opposed to a cloud provider specific integration for this where we allow the cloud provider to do initialization in 2 steps - one for node addresses, topology, etc and another for node readiness - but I would rather have a generic solution that fixes this for many other use-cases.